### PR TITLE
start to freeze dependencies for docs

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
Now that we have a workflow that freezes deps when running tests, we should have one that freezes deps when building docs too! This is almost a copy-paste of [how tests create freeze files](https://github.com/PennyLaneAI/pennylane/blob/f7bba2b8a21319309b21250073492573f5ee967a/.github/workflows/tests.yml#L42-L91) on master today, but with some tweaks.

**Description of the Change:**
Run `pip freeze` right before building the docs, then push it to (or push and open) a PR to maintain the stable results on GitHub. Some notes:
- There was a missing `&&` so this action is not currently installing proper environments. I fixed it here
- unlike tests, I save the results to a file with the `.tmp` suffix. This is because I have to create it using the `pre-build-command` on the user's branch, whereas I am able to check out the deps branch before on the test suite. Subtle difference, but I had conflicts without it. Praise be `mv -f`

For example, check out the commit made to an existing PR [here](https://github.com/PennyLaneAI/pennylane/pull/5210/commits/77428e0fafe99ce3decf8bc0bf416186bdcca1bc)

**Benefits:**
- We will have a record of stable dependencies for building docs
- The docs environment is now correctly configured for this check

**Possible Drawbacks:**
- this runs on PRs, and we didn't want that to happen for tests. I'm only like 70% sure that there won't be unexpected errors because of it
- I was considering removing this action in #5136 altogether, and only having one docs build run (yep, we build it twice!). the context for why is at the top of `docs.yml`, but I was hoping to fix it. Unfortunately (as you can see in that PR), I can't make this change within the RTD build because it doesn't have write-access to GitHub. If anyone has a clue of how to do it there instead (maybe rsync to somewhere else, then have that action commit the change?), please lmk
- if two PRs run this simultaneously, there might be a race condition, and that would probably make the latter PR's CI fail due to the branch being out-of-date with the remote. they'd have to finish the "Documentation Check" action within a few seconds of each other AND there be an outdated dependency for this to occur, so it seems rather unlikely

### CODE REVIEW

**Review by rashidnhm — APPROVED** [review]

**Comment by rashidnhm** [comment]
Should this push to a different branch, do we have potential of race conditions?

**Comment by timmysilv** [comment]
there's a small chance it'll conflict with the commit triggered by tests, but it'd be down to the wire if it happens. imo I wouldn't worry about it because I enjoy only having one dependencies PR/branch to keep track of, but if we're scared we can consider it

File: .github/workflows/docs.yml [hunk 1]
```diff
@@ -18,13 +18,64 @@ name: "Documentation check"
 on:
 - pull_request
 
+env:
+  DEPS_BRANCH: bot/stable-deps-update
```
Comment (Mandrenkov) [inline .github/workflows/docs.yml:22]: # Question

Why not define this as an [environment variable of the job](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idenv) to keep things more tightly scoped?

File: .github/workflows/docs.yml [hunk 2]
```diff
@@ -18,13 +18,64 @@ name: "Documentation check"
...
+
 jobs:
   sphinx:
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v1
+    - uses: actions/checkout@v3
```
Comment (Mandrenkov) [inline .github/workflows/docs.yml:28]: # Question

Any particular reason we are not using [v4](https://github.com/actions/checkout)?
```suggestion
    - uses: actions/checkout@v4
```

File: .github/workflows/docs.yml [hunk 3]
```diff
@@ -18,13 +18,64 @@ name: "Documentation check"
...
     - uses: PennyLaneAI/sphinx-action@master
       with:
         docs-folder: "doc/"
-        pre-build-command: "apt install -y graphviz && pip3 install -r doc/requirements.txt && pip3 install . pip3 install openfermionpyscf && pip3 install aiohttp fsspec h5py"
+        pre-build-command: "apt install -y graphviz && pip3 install -r doc/requirements.txt && pip3 install . && pip3 install openfermionpyscf && pip3 install aiohttp fsspec h5py && pip freeze | grep -v 'file:///' > .github/stable/doc.txt.tmp"
```
Comment (Mandrenkov) [inline .github/workflows/docs.yml:32]: # Suggestion
Fortunately, YAML can help us with readability:
```suggestion
        pre-build-command: >
            apt install -y graphviz
            && pip3 install -r doc/requirements.txt
            && pip3 install .
            && pip3 install openfermionpyscf
            && pip3 install aiohttp fsspec h5py
            && pip freeze | grep -v 'file:///' > .github/stable/doc.txt.tmp
```

File: .github/workflows/docs.yml [hunk 4]
```diff
@@ -18,13 +18,64 @@ name: "Documentation check"
...
+          Automatic update of stable requirement files to snapshot valid python environments.
+
+          Because bots are not able to trigger CI on their own, please do so by pushing an empty commit to this branch using the following command:
+
+          ```
+          git commit --allow-empty -m 'trigger ci'
+          ```
+
+          Alternatively, wait for this branch to be out-of-date with master, then just use the "Update branch" button!
+        pr_allow_empty: false
```
Comment (Mandrenkov) [inline .github/workflows/docs.yml:80]: # Question

Does this mean we don't actually need `if: steps.changed.outputs.has_changes != '0'` above?

File: .github/workflows/docs.yml [hunk 5]
```diff
@@ -18,13 +18,64 @@ name: "Documentation check"
...
+      if: steps.changed.outputs.has_changes != '0'
+      run: |
+        git add .github/stable/doc.txt
+        git commit -m "Update stable docs dependencies"
+        git push -f --set-upstream origin "${{ env.DEPS_BRANCH }}"
+
+    # Create PR to master
+    - name: Create pull request
+      if: steps.changed.outputs.has_changes != '0'
+      uses: repo-sync/pull-request@v2
```
Comment (Mandrenkov) [inline .github/workflows/docs.yml:64]: # Thought

This [action](https://github.com/repo-sync/pull-request) has now been archived in favour of using the GitHub CLI. 

Should we (eventually) migrate to using the GitHub CLI?

File: .github/workflows/docs.yml [hunk 6]
```diff
@@ -18,13 +18,64 @@ name: "Documentation check"
...
+
+    - name: Stage changes
+      if: steps.changed.outputs.has_changes != '0'
+      run: |
+        git add .github/stable/doc.txt
+        git commit -m "Update stable docs dependencies"
+        git push -f --set-upstream origin "${{ env.DEPS_BRANCH }}"
+
+    # Create PR to master
+    - name: Create pull request
```
Comment (Mandrenkov) [inline .github/workflows/docs.yml:62]: # Question

I missed this last time, but what happens if a PR already exists? Nothing happens?

**Review by Mandrenkov — COMMENTED** [review]
Overall, looks good to me! 🥇

File: .github/workflows/docs.yml (hunk 1 above)
Comment (timmysilv) [reply .github/workflows/docs.yml:22]: no reason, I'll try it out

File: .github/workflows/docs.yml (hunk 4 above)
Comment (timmysilv) [reply .github/workflows/docs.yml:86]: probably but I'm happy with the double-safeguard. just skip this step if we don't need it, and if somehow there are no commits despite us expecting one, we'll get a nice error here

File: .github/workflows/docs.yml (hunk 2 above)
Comment (Mandrenkov) [reply .github/workflows/docs.yml:27]: Sure, sounds reasonable to me!

File: .github/workflows/docs.yml (hunk 4 above)
Comment (Mandrenkov) [reply .github/workflows/docs.yml:85]: Gotcha. I was mainly asking to make sure I understood things correctly; this sounds good to me.

File: .github/workflows/docs.yml (hunk 5 above)
Comment (Mandrenkov) [reply .github/workflows/docs.yml:69]: It wouldn't be a bad idea to capture these CI upgrades in a Story to be tackled sometime later!

**Review by Mandrenkov — APPROVED** [review]

**Review by albi3ro — APPROVED** [review]
Seems straightforward enough. Thanks for looking into all of this :)

**Pull request merged**